lib/client/keystore.go

```diff
@@ -451,33 +449,65 @@ func (o withDBCerts) deleteKey(dirPath, username string) error {
 	return os.RemoveAll(filepath.Join(dirPath, username+dbDirSuffix, o.teleportClusterName))
 }
 
-// SaveCerts saves trusted TLS certificates of certificate authorities
-func (fs *FSLocalKeyStore) SaveCerts(proxy string, cas []auth.TrustedCerts) (retErr error) {
-	dir, err := fs.dirFor(proxy, true)
-	if err != nil {
-		return trace.Wrap(err)
+// initKeysDir initializes the keystore root directory. Usually it is ~/.tsh
+func initKeysDir(dirPath string) (string, error) {
+	var err error
+	// not specified? use `~/.tsh`
+	if dirPath == "" {
+		u, err := user.Current()
+		if err != nil {
+			dirPath = os.TempDir()
+		} else {
+			dirPath = u.HomeDir
+		}
+		dirPath = filepath.Join(dirPath, defaultKeyDir)
 	}
-	fp, err := os.OpenFile(filepath.Join(dir, fileNameTLSCerts), os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0640)
+	// create if doesn't exist:
+	_, err = os.Stat(dirPath)
 	if err != nil {
-		return trace.ConvertSystemError(err)
-	}
-	defer utils.StoreErrorOf(fp.Close, &retErr)
-	for _, ca := range cas {
-		for _, cert := range ca.TLSCertificates {
-			if _, err := fp.Write(cert); err != nil {
-				return trace.ConvertSystemError(err)
-			}
-			if _, err := fmt.Fprintln(fp); err != nil {
-				return trace.ConvertSystemError(err)
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(dirPath, os.ModeDir|profileDirPerms)
+			if err != nil {
+				return "", trace.ConvertSystemError(err)
 			}
+		} else {
+			return "", trace.Wrap(err)
 		}
 	}
-	return fp.Sync()
+
+	return dirPath, nil
+}
+
+type fsLocalNonSessionKeyStore struct {
+	// log holds the structured logger.
+	log *logrus.Entry
+
+	// KeyDir is the directory where all keys are stored.
+	KeyDir string
+}
+
+// dirFor returns the path to the session keys for a given host. The value
+// for fs.KeyDir is typically "~/.tsh", sessionKeyDir is typically "keys",
+// and proxyHost typically has values like "proxy.example.com".
+//
+// If the create flag is true, the directory will be created if it does
+// not exist.
+func (fs *fsLocalNonSessionKeyStore) dirFor(proxyHost string, create bool) (string, error) {
```

Comment:
Using a boolean saves a couple of keystrokes at the cost of obscuring the flow. Create directory where necessary explicitly:
```go
func (fs *fsLocalNonSessionKeyStore) SaveCerts(proxy string, cas []auth.TrustedCerts) (retErr error) {
	dir := fs.dirFor(proxy)
	if err := os.MkdirAll(dir, profileDirPerms); err != nil {
		fs.log.Error(err)
		return trace.ConvertSystemError(err)
	}
	// ...
}
```